### PR TITLE
cri: Add termination_grace_period for Pod Sandbox

### LINKED ISSUE #77873: Stop timeout isn't respected at shutdown/reboot
**What happened**:
Containers are getting terminated by systemd without respecting the `terminationGracePeriodSeconds` set in the pod yaml on reboot or shutdown of a node.

**What you expected to happen**:
`terminationGracePeriodSeconds` is respected by systemd when using systemd as the cgroup manager. 

**How to reproduce it (as minimally and precisely as possible)**:
1. Use systemd as the cgroup manager in your container runtime.
1. Create a pod yaml with `terminationGracePeriodSeconds` set to `120` seconds.
1. Reboot the node.
1. You will notice that the containers get SIGTERM, followed by systemd default stop timeout (typically 90 seconds) and then they are SIGKILLed.

**Anything else we need to know?**:
This can be fixed by passing the stop timeout to the containers as part of the CreateContainer CRI API. This will allow the container runtimes to set the systemd property for the scope to override the default stop timeout to the value set through `terminationGracePeriodSeconds`.
This needs changes across the stack as runc doesn't currently provide a way to set the TimeoutStopUSec for the systemd scope created for a container.
The behavior with `cgroupfs` cgroup manager will need further investigation.

**Environment**:
- Kubernetes version (use `kubectl version`): All versions.

### PR DESCRIPTION
**Pull request by mrunalp**
**What type of PR is this?**:
/kind bug

**What this PR does / why we need it**:
This PR adds termination_grace_period to the Pod Sandbox in CRI API to handle
system shutdowns/reboots gracefully for the pod containers.
One way to achieve it is by setting the systemd TimeoutStopUSec property when
using the systemd cgroup manager. 
See #77873 for more details.

**Which issue(s) this PR fixes**:

Fixes #77873

```release-note
Add termination_grace_period to PodSandboxConfig
```
/sig node
/priority important-soon
/cc @derekwaynecarr
/assign @yujuhong
/milestone v1.15

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
@mrunalp thank you for the summary and issue description as it captures what we have discovered in practice well.  depending on the workload on the host, not respecting graceful shutdown is unexpected in response to normal power-off.

@PatrickLang what behavior do you see in windows when you shutdown the host?  is the pod specific termination graceperiod respected?

@yujuhong or @dashpole are you able to see what behaviors we get with cgroupfs?  have you seen similar issues in your environments?

**Comment by derekwaynecarr** [comment]
I would like to address this bug in 1.15 if possible, but understand that a change to CRI has broader impact so I want to respect input from wide audience.

**Comment by PatrickLang** [comment]
Following the breadcrumbs - it looks like this should be possible on Windows if it's set up in dockershim. However, I don't see that implemented for Linux or Windows in this PR.


Windows references:
https://github.com/kubernetes/kubernetes/issues/73434  - this tracks the failing tests for Windows
https://github.com/moby/moby/issues/25982 - related moby issue, which should have been resolved from an API standpoint in https://github.com/moby/moby/pull/28563

**Comment by mrunalp** [comment]
@PatrickLang This will have to be set in runc to work in Linux (at least for the systemd cgroup manager case). I have a runc PR open here: https://github.com/opencontainers/runc/pull/2062. This PR adds the CRI changes and the client changes, so the CRI implementations can implement this as they see fit. For CRI-O, we will use the runc change as we default to the systemd cgroup manager.

**Comment by PatrickLang** [comment]
I'm ok with the API change, but don't have a place to implement it yet as CRI-ContainerD isn't ready.

Will use https://github.com/kubernetes/kubernetes/issues/73434 to track this for Windows support

**Comment by PatrickLang** [comment]
Actually, I misunderstood this change. I just talked to @derekwaynecarr about the subtlety here of this being a config property that should be applied even if StopPodSandbox isn't called. Still digging

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 1]
```diff
@@ -357,6 +357,13 @@ message PodSandboxConfig {
     map<string, string> annotations = 7;
     // Optional configurations specific to Linux hosts.
     LinuxPodSandboxConfig linux = 8;
+    // termination_grace_period is the time that a pod reaper should wait
+    // before terminating the containers in a pod.
+    // For e.g., on a Linux system, this can be used to configure the
+    // systemd TimeoutStopUSec property for the containers running in systemd
+    // scopes, so systemd gives the pod containers enough time to shutdown
+    // gracefully on system shutdown or reboot.
+    int64 termination_grace_period = 9;
```
Comment (PatrickLang) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: @jterry75 brings up a good point - why is this called `termination_grace_period` if it's for shutdown?

**Comment by derekwaynecarr** [comment]
The desire is to ensure that pods that are not typically drained (static pods or daemonsets) have their termination period respected.  It is also our desire to ensure that even if the kubelet was not healthy/running on a host that the runtime could configure the sandbox with the proper termination period in advance such that power down of a host does not require interactions from the kubelet to ensure the stop period is respected.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 1 above)
Comment (PatrickLang) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: @egernst - does this make sense for Kata?

Comment (derekwaynecarr) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Meant to align with the Pod terminationGracePeriodSeconds field

**Comment by PatrickLang** [comment]
Ok, more info - this would still need to be implemented in ContainerD on Windows if a configurable timeout is needed on shutdown without an explicit StopPodSandbox call

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 1 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: This is for graceful termination at either shutdown or reboot, so I feel that `termination_grace_period` covers both of those cases in the name.

**Review by mattjmcnaughton — COMMENTED** [review]
Looks like there's a lot of good thoughts here already :)

The only point I'd like to add is whether it'd be possible/advisable for us to add some form of integration/e2e testing here?

**Comment by yujuhong** [comment]
@derekwaynecarr @mrunalp I think the feature can be useful, but I don't fully understand the use cases, caveats, and alternative approaches. I think this warrants more discussion. WDYT?

/hold

**Comment by Random-Liu** [comment]
Can we get the runc side discussion done and change in first?

We need to make sure we have a clear use case and viable implementation before changing CRI. Or else, how do we test the CRI change? And how do we know the CRI change make sense?

**Pull request closed without merging**